python/ray/_raylet.pyx

```diff
@@ -1382,43 +1388,69 @@ async def execute_streaming_generator_async(
 
     gen = context.generator
 
-    futures = []
-    while True:
-        try:
-            output_or_exception = await gen.__anext__()
-        except StopAsyncIteration:
-            break
-        except AsyncioActorExit:
-            # The execute_task will handle this case.
-            raise
-        except Exception as e:
-            output_or_exception = e
+    loop = asyncio.get_running_loop()
 
-        loop = asyncio.get_running_loop()
-        worker = ray._private.worker.global_worker
+    worker = ray._private.worker.global_worker
+    executor = worker.core_worker.get_event_loop_executor()
+
+    interrupt_signal_event = threading.Event()
+    futures = []
 
-        # NOTE: Reporting generator output in a streaming fashion,
-        #       is done in a standalone thread-pool fully *asynchronously*
-        #       to avoid blocking the event-loop and allow it to *concurrently*
-        #       make progress, since serializing and actual RPC I/O is done
-        #       with "nogil".
-        futures.append(
-            loop.run_in_executor(
-                worker.core_worker.get_thread_pool_for_async_event_loop(),
-                report_streaming_generator_output,
-                context,
-                output_or_exception,
-                cur_generator_index,
+    try:
+        while True:
+            try:
+                output_or_exception = await gen.__anext__()
+            except StopAsyncIteration:
+                break
+            except AsyncioActorExit:
+                # The execute_task will handle this case.
+                raise
+            except Exception as e:
+                output_or_exception = e
+
+            # NOTE: Reporting generator output in a streaming fashion,
+            #       is done in a standalone thread-pool fully *asynchronously*
+            #       to avoid blocking the event-loop and allow it to *concurrently*
+            #       make progress, since serializing and actual RPC I/O is done
+            #       with "nogil".
+            futures.append(
+                loop.run_in_executor(
+                    executor,
+                    report_streaming_generator_output,
+                    context,
+                    output_or_exception,
+                    cur_generator_index,
+                    interrupt_signal_event,
+                )
             )
-        )
 
-        cur_generator_index += 1
+            cur_generator_index += 1
 
-        if isinstance(output_or_exception, Exception):
-            break
+            if isinstance(output_or_exception, Exception):
+                break
+
+        # Make sure all RPC I/O completes before returning
+        await asyncio.gather(*futures)
```

Comment:
Q: isn't using await asyncio.gather(*futures, return_exceptions=True) solve the issue?

Reply:
Answered here: https://github.com/ray-project/ray/pull/43775#issuecomment-1986208494